[NCCL] Dedicated stream to run all FutureNCCL callbacks.

sinannasir opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #43310 DDP communication hook examples
* **#43447 [NCCL] Dedicated stream to run all FutureNCCL callbacks.**

Two main better-engineering motivations to run all FutureNCCL callbacks on a dedicated stream:
1. Each time a then callback was called, we would get a stream from the pool and run the callback on that stream. If we observe the stream traces using that approach, we would see a lot of streams and debugging would become more complicated. If we have a dedicated stream to run all then callback operations, the trace results will be much cleaner and easier to follow.
2. StreamPool may eventually return the default stream or a stream that is used for other operations. This can cause slowdowns.

Unless then callback takes longer than preceding allreduce, this approach will be as performant as the previous approach.

Differential Revision: [D23277575](https://our.internmc.facebook.com/intern/diff/D23277575/)

sinannasir committed:
[NCCL] Dedicated stream to run all FutureNCCL callbacks.

Two main better-engineering motivations to run all FutureNCCL callbacks on a dedicated stream:
1. Each time a then callback was called, we would get a stream from the pool and run the callback on that stream. If we observe the stream traces using that approach, we would see a lot of streams and debugging would become more complicated. If we have a dedicated stream to run all then callback operations, the trace results will be much cleaner and easier to follow.
2. StreamPool may eventually return the default stream or a stream that is used for other operations. This can cause slowdowns.

Unless then callback takes longer than preceding allreduce, this approach will be as performant as the previous approach.

Differential Revision: [D23277575](https://our.internmc.facebook.com/intern/diff/D23277575/)

[ghstack-poisoned]

sinannasir committed:
Update on "[NCCL] Dedicated stream to run all FutureNCCL callbacks."


Two main better-engineering motivations to run all FutureNCCL callbacks on a dedicated stream:
1. Each time a then callback was called, we would get a stream from the pool and run the callback on that stream. If we observe the stream traces using that approach, we would see a lot of streams and debugging would become more complicated. If we have a dedicated stream to run all then callback operations, the trace results will be much cleaner and easier to follow.
2. StreamPool may eventually return the default stream or a stream that is used for other operations. This can cause slowdowns.

Unless then callback takes longer than preceding allreduce, this approach will be as performant as the previous approach.

Differential Revision: [D23277575](https://our.internmc.facebook.com/intern/diff/D23277575/)

[ghstack-poisoned]

pritamdamania87 commented on torch/lib/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -644,6 +644,13 @@ std::vector<std::shared_ptr<NCCLComm>>& ProcessGroupNCCL::getNCCLComm(
+
+    // If single-process single-device mode, WorkNCCL::getFuture is supported,
+    // so get a dedicated stream to run FutureNCCL then callbacks.
+    if (devices.size() == 1) {
+      futureNCCLCallbackStream_ =
+          std::make_shared<at::cuda::CUDAStream>(at::cuda::getStreamFromPool());
+    }
```
This should be done in the constructor of ProcessGroupNCCL and not each time we call `getNCCLComm`.

sinannasir replied:
I've decided to make a static vector of streams called `futureNCCLCallbackStreams_` instead that will store one stream per device. I think it will look way better. I'll give a quick try and let you know if it causes problems.

sinannasir replied:
That's a bit strange but when I try to initialize even a simple int using c10::cuda::device_count: `int ProcessGroupNCCL::test = c10::cuda::device_count();`; I get the following error:
```
THCudaCheck FAIL file=../aten/src/THC/THCGeneral.cpp line=54 error=3 : initialization error
ERROR:root:Caught exception: 
Traceback (most recent call last):
  File "/home/sinannasir/local/pytorch/torch/testing/_internal/common_distributed.py", line 226, in wrapper
    fn()
  File "/home/sinannasir/local/pytorch/torch/testing/_internal/common_distributed.py", line 93, in wrapper
    return func(*args, **kwargs)
  File "test/distributed/test_c10d.py", line 3226, in test_ddp_comm_hook_allreduce_with_then_hook_nccl
    gpu_model = self._gpu_model_with_ddp_comm_hook(
  File "test/distributed/test_c10d.py", line 3111, in _gpu_model_with_ddp_comm_hook
    TestDdpCommHook().to(device_id),
  File "/home/sinannasir/local/pytorch/torch/nn/modules/module.py", line 612, in to
    return self._apply(convert)
  File "/home/sinannasir/local/pytorch/torch/nn/modules/module.py", line 359, in _apply
    module._apply(fn)
  File "/home/sinannasir/local/pytorch/torch/nn/modules/module.py", line 381, in _apply
    param_applied = fn(param)
  File "/home/sinannasir/local/pytorch/torch/nn/modules/module.py", line 610, in convert
    return t.to(device, dtype if t.is_floating_point() else None, non_blocking)
  File "/home/sinannasir/local/pytorch/torch/cuda/__init__.py", line 173, in _lazy_init
    torch._C._cuda_init()
RuntimeError: cuda runtime error (3) : initialization error at ../aten/src/THC/THCGeneral.cpp:54
exiting process with exit code: 10
THCudaCheck FAIL file=../aten/src/THC/THCGeneral.cpp line=54 error=3 : initialization error
ERROR:root:Caught exception: 
Traceback (most recent call last):
  File "/home/sinannasir/local/pytorch/torch/testing/_internal/common_distributed.py", line 226, in wrapper
    fn()
  File "/home/sinannasir/local/pytorch/torch/testing/_internal/common_distributed.py", line 93, in wrapper
    return func(*args, **kwargs)
  File "test/distributed/test_c10d.py", line 3226, in test_ddp_comm_hook_allreduce_with_then_hook_nccl
    gpu_model = self._gpu_model_with_ddp_comm_hook(
  File "test/distributed/test_c10d.py", line 3111, in _gpu_model_with_ddp_comm_hook
    TestDdpCommHook().to(device_id),
  File "/home/sinannasir/local/pytorch/torch/nn/modules/module.py", line 612, in to
    return self._apply(convert)
  File "/home/sinannasir/local/pytorch/torch/nn/modules/module.py", line 359, in _apply
    module._apply(fn)
  File "/home/sinannasir/local/pytorch/torch/nn/modules/module.py", line 381, in _apply
    param_applied = fn(param)
  File "/home/sinannasir/local/pytorch/torch/nn/modules/module.py", line 610, in convert
    return t.to(device, dtype if t.is_floating_point() else None, non_blocking)
  File "/home/sinannasir/local/pytorch/torch/cuda/__init__.py", line 173, in _lazy_init
    torch._C._cuda_init()
RuntimeError: cuda runtime error (3) : initialization error at ../aten/src/THC/THCGeneral.cpp:54
```

I just initialize the streams inside constructor and they are not static.

pritamdamania87 commented on torch/lib/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -799,8 +809,13 @@ std::shared_ptr<ProcessGroup::Work> ProcessGroupNCCL::collective(
 
-  // Store a reference to outputs to be used by WorkNCCL::getFuture.
-  work->outputs_ = std::make_shared<std::vector<at::Tensor>>(outputs);
+  // If single-process single-device mode, WorkNCCL::getFuture is supported.
+  // Store references to outputs and futureNCCLCallbackStream to be used by
+  // WorkNCCL::getFuture.
+  if (outputs.size() == 1) {
```
I think it's fine to always do this instead of adding a condition here. We already have the appropriate check in FutureNCCL for this right?

sinannasir replied:
After defining FutureNCCL callback streams inside the constructor. I guess if I remove that check, the following will cause linter warning:
https://github.com/pytorch/pytorch/blob/d9c8b2e4613209f5f354df121fff61dadfc93afe/torch/lib/c10d/ProcessGroupNCCL.cpp#L828-L829

pritamdamania87 replied:
What sort of linter error do you get? We should probably pass all the futureNCCLCallbackStreams_ to `WorkNCCL` as well and not just one. That way in the future it's easier to extend support for single process multi device mode.

sinannasir replied:
Agreed, I thought that if I do `outputs[0]` without any size check it would give a linter error. If we just pass a vector, it should be okay.

pritamdamania87 commented on torch/lib/c10d/ProcessGroupNCCL.hpp:
```diff
@@ -237,12 +246,9 @@ class ProcessGroupNCCL : public ProcessGroup {
         std::function<void(void)> callback,
-        const c10::cuda::CUDAStream& stream,
-        std::shared_ptr<std::vector<at::cuda::CUDAEvent>> thenFutCudaEvents) {
+        const c10::cuda::CUDAStream& stream) {
```
We don't need to explicitly pass the stream in here right? We can just use `futureNCCLCallbackStream_` directly in the method?

pritamdamania87 reviewed (approved):
LGTM, please address some of my comments below before landing.

pritamdamania87 commented on torch/lib/c10d/ProcessGroupNCCL.hpp:
```diff
@@ -237,12 +246,9 @@ class ProcessGroupNCCL : public ProcessGroup {
-      (*thenFutCudaEvents)[0].record(stream);
     }
 
     // We use addCallbackWithStream instead of addCallback.
```
We can just use `addCallback` here instead of `addCallbackWithStream` since we no longer need to pass the stream?

sinannasir committed:
Update on "[NCCL] Dedicated stream to run all FutureNCCL callbacks."


Two main better-engineering motivations to run all FutureNCCL callbacks on a dedicated stream:
1. Each time a then callback was called, we would get a stream from the pool and run the callback on that stream. If we observe the stream traces using that approach, we would see a lot of streams and debugging would become more complicated. If we have a dedicated stream to run all then callback operations, the trace results will be much cleaner and easier to follow.
2. StreamPool may eventually return the default stream or a stream that is used for other operations. This can cause slowdowns.

Unless then callback takes longer than preceding allreduce, this approach will be as performant as the previous approach.

Differential Revision: [D23277575](https://our.internmc.facebook.com/intern/diff/D23277575/)

[ghstack-poisoned]

sinannasir committed:
Update on "[NCCL] Dedicated stream to run all FutureNCCL callbacks."


Two main better-engineering motivations to run all FutureNCCL callbacks on a dedicated stream:
1. Each time a then callback was called, we would get a stream from the pool and run the callback on that stream. If we observe the stream traces using that approach, we would see a lot of streams and debugging would become more complicated. If we have a dedicated stream to run all then callback operations, the trace results will be much cleaner and easier to follow.
2. StreamPool may eventually return the default stream or a stream that is used for other operations. This can cause slowdowns.

Unless then callback takes longer than preceding allreduce, this approach will be as performant as the previous approach.

Differential Revision: [D23277575](https://our.internmc.facebook.com/intern/diff/D23277575/)

[ghstack-poisoned]

sinannasir committed:
Update on "[NCCL] Dedicated stream to run all FutureNCCL callbacks."


Two main better-engineering motivations to run all FutureNCCL callbacks on a dedicated stream:
1. Each time a then callback was called, we would get a stream from the pool and run the callback on that stream. If we observe the stream traces using that approach, we would see a lot of streams and debugging would become more complicated. If we have a dedicated stream to run all then callback operations, the trace results will be much cleaner and easier to follow.
2. StreamPool may eventually return the default stream or a stream that is used for other operations. This can cause slowdowns.

Unless then callback takes longer than preceding allreduce, this approach will be as performant as the previous approach.

Differential Revision: [D23277575](https://our.internmc.facebook.com/intern/diff/D23277575/)

[ghstack-poisoned]

pritamdamania87 commented on torch/lib/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -409,6 +409,24 @@ ProcessGroupNCCL::ProcessGroupNCCL(
+  // Depending on the device index of collective outputs, WorkNCCL passes
+  // the corresponding device's then callback stream to FutureNCCL.
+  futureNCCLCallbackStreams_.reserve(c10::cuda::device_count());
+  try {
+    for (int device_index = 0; device_index < c10::cuda::device_count();
+         device_index++) {
+      futureNCCLCallbackStreams_.push_back(
+          std::make_shared<at::cuda::CUDAStream>(
+              at::cuda::getStreamFromPool(device_index)));
+    }
+  } catch (const std::exception& e) {
+    throw std::runtime_error(
+        "Error when inititalizing futureNCCLCallbackStreams: " +
+        std::string(e.what()));
+  }
```
Why do we have a try-catch here which basically just rethrows the exception? We can just avoid having a try-catch completely and let the original exception propagate?

sinannasir replied:
I was getting unhandled exception when trying the static int that I mentioned on top:
https://github.com/pytorch/pytorch/pull/43447#discussion_r478490040

That's why I put try-catch there.

pritamdamania87 commented on torch/lib/c10d/ProcessGroupNCCL.hpp:
```diff
@@ -136,6 +136,8 @@ class ProcessGroupNCCL : public ProcessGroup {
 
     // Store a reference to NCCL collective's outputs to be used by getFuture.
     std::shared_ptr<std::vector<at::Tensor>> outputs_;
+    // Store a reference to the NCCL stream that runs all FutureNCCL callbacks.
+    std::shared_ptr<at::cuda::CUDAStream> futureNCCLCallbackStream_;
```
Pass a vector here for all the streams for all devices. We can have the limitations for single process multi device mode only in FutureNCCL where its necessary.

sinannasir committed:
Update on "[NCCL] Dedicated stream to run all FutureNCCL callbacks."


Two main better-engineering motivations to run all FutureNCCL callbacks on a dedicated stream:
1. Each time a then callback was called, we would get a stream from the pool and run the callback on that stream. If we observe the stream traces using that approach, we would see a lot of streams and debugging would become more complicated. If we have a dedicated stream to run all then callback operations, the trace results will be much cleaner and easier to follow.
2. StreamPool may eventually return the default stream or a stream that is used for other operations. This can cause slowdowns.

Unless then callback takes longer than preceding allreduce, this approach will be as performant as the previous approach.

Differential Revision: [D23277575](https://our.internmc.facebook.com/intern/diff/D23277575/)

[ghstack-poisoned]

pritamdamania87 reviewed (approved):
LGTM!

sinannasir committed:
Update on "[NCCL] Dedicated stream to run all FutureNCCL callbacks."


Two main better-engineering motivations to run all FutureNCCL callbacks on a dedicated stream:
1. Each time a then callback was called, we would get a stream from the pool and run the callback on that stream. If we observe the stream traces using that approach, we would see a lot of streams and debugging would become more complicated. If we have a dedicated stream to run all then callback operations, the trace results will be much cleaner and easier to follow.
2. StreamPool may eventually return the default stream or a stream that is used for other operations. This can cause slowdowns.

Unless then callback takes longer than preceding allreduce, this approach will be as performant as the previous approach.

Differential Revision: [D23277575](https://our.internmc.facebook.com/intern/diff/D23277575/)

[ghstack-poisoned]
